### DISCUSSION TITLE
I want to change from Hyperion to HyperHDR.

### DISCUSSION THREAD

**Discussion by bong8839** [awawa-dev/HyperHDR#182] [Q&A]
Hello. I'm a Korean.
Please understand that the grammar may be incorrect due to the necessity of using a translator to write the text.

Currently building an ambient light using Hyperion + Pi4 + Arduino uno + ws2812b 6m 
However, there seems to be a little delay and a problem with black bar detection, so I want to change to your system, but the problem is that I am a very newbie.

First of all, my system is structured like this,
![image](https://user-images.githubusercontent.com/97230290/148399256-fb2b27d1-57ab-44be-9c62-bdaf611b1713.png)

And the Arduino Uno and LED wiring are configured in this way.
![34b2455baa55e0](https://user-images.githubusercontent.com/97230290/148399417-fdf36de6-2c82-466c-8742-66e1ba4fc059.jpg)

Reading your system configuration article, I was told to purchase and configure ESP8266 or ESP32 instead of Arduino.

1. Which esp32 or 8266 is a better choice?
2. Below is the 8266 board, how should I wire it with the LED data, GND pin?
![image](https://user-images.githubusercontent.com/97230290/148400514-348d6b18-6b16-4085-9060-f5abc7960cd8.png)

Sorry for such a novice question. Your system compensates for many of the shortcomings of Hyperion and I want to change it, but no one in Korea knows your system.

Please help.

**Comment by awawa-dev** [comment]
Hi
HyperHDR split from Hyperion quite long ago, we left some of its functionalities and included more new features, bug fixes & improvements. Still the migration should be very easy if you know the other system. There are also some tutorials in the wiki and on the blog.

> ws2812b 6m

Well, that's a lot. Don't know the density, but for 60led/meter its nearly impossible to make it work properly using Arduino. It's too slow.

I can propose 2 solutions:

- if you can put your Rpi close to the Esp32/Esp8266 and the LED strip, then chose HyperSPI and generic Esp32 connected with Rpi using short SPI jumper cables. Esp32 is a bit better because it can handle such number of LEDs in one large SPI frame and Esp8266 needs to receive large number of smaller SPI frame first and then combine them. That's the best solution, but it depends on the placement of the components .

- if you want to use Esp32/Esp8266 just like Arduino connected with the USB cable and using the Adalight high speed 2Mb AWA protocol,  then it could be tricky because Esp32/Esp8266 needs to have 2Mb speed capable serial port chip onboard like ch340g, cp2104, CH9102x. But there are a lot of counterfeits on the market recently, ch340g is affected especially....I recommend to give it a try for Esp32 CH9102x (sample offer you can find in this thread: [link](https://github.com/awawa-dev/HyperHDR/issues/129))  as it's universal device and later you can choose between HyperSerialEsp32/HyperSerialWLED and HyperSPI depending on your needs.

If you are using the USB grabber and Adalight USB serial solution simultaneously with Rpi, you may face the problem of USB Rpi limited performance. Probably the MS2109 clone is not affected as it offers a rather low video stream resolution and frame rate. You can see the problem in the screenshots here: https://github.com/awawa-dev/HyperSPI 

**Connecting the RPI to the LED strip power supply almost always leads to a problem (more or less serious) with the stability of Rpi**
 It is recommended to use a second solid 2A USB power supply with capable USB power cable for Rpi and provide a "common" ground between the LED power supply (minus) and Rpi (GPIO ground). 

**You need also 3.3V to 5V level shifter like SN74AHCT125N (example https://github.com/awawa-dev/HyperHDR/discussions/177 )**

**Reply by bong8839** [reply]
Thank you for answer. The density is 1M/60LED, and the total number of LEDs is 350 using less than 6M.

It's a working video, but it's not very slow, but there's a very slight delay.
https://www.youtube.com/watch?v=mVGrtHEbn1o

The first solution said the distance should be close, my system Rpi4 + esp8266 or esp32 can be installed right next to it and the LED + - data lines can be connected with wires within about 1M.

At this distance, can I try the first solution?
If possible, can I just prepare esp8266 and wire?
Ah. Level shifter SN74AHCT125N Do I need to prepare this as well?

Are you saying that you need to prepare a separate power supply for Rpi4? And do you connect the gpio ground terminal to the SMPS minus that powers the LEDs?


I roughly drew a wiring diagram, but for the SN74AHCT125N connection,
 I don't know about it.. Could you please check the wiring easily?
![image](https://user-images.githubusercontent.com/97230290/148423650-d7925c7d-a95b-4fe0-b0da-5d06a1d87103.png)

**Comment by awawa-dev** [comment]
> Thank you for answer. The density is 1M/60LED, and the total number of LEDs is 350 using less than 6M.

Maximum what you can achieve using proper solution is around ~90Hz LED refresh rate. But Arduino is not capable to handle it...it's explained here: https://github.com/awawa-dev/HyperSerialEsp8266 Probably you have around 12-13Hz currently.

> It's a working video, but it's not very slow, but there's a very slight delay. https://www.youtube.com/watch?v=mVGrtHEbn1o

That delay is rather low. Did you try to tune-up configuration? Please follow the "performance pro-tip" section in the introduction.
https://github.com/awawa-dev/HyperHDR/wiki/Quick-start
Of course there is a lot of room for improvement on hardware level but sometimes we can improve the situation a bit with the proper configuration alone. But it won't improve your limited refresh rate because it's a hardware limitation.

> The first solution said the distance should be close, my system Rpi4 + esp8266 or esp32 can be installed right next to it and the LED + - data lines can be connected with wires within about 1M.

OK. If the Esp could be install next to Rpi then HyperSPI can be used. But it's rather a long distance (1meter) for data cable. The EMI interference is one problem. The second is the voltage drop on data line. Arduino provides logic at 5v and it could work, but Rpi and esp are 3.3v...sometimes  it works also but rather not for that distance due to the voltage drop. Read more here about 3.3V/5V subject: https://hyperhdr.blogspot.com/2020/12/my-build-log-using-sk6812-rgbw-led.html#chapter3

> At this distance, can I try the first solution? If possible, can I just prepare esp8266 and wire? Ah. Level shifter SN74AHCT125N Do I need to prepare this as well?

My cables between Rpi and Esp are 15cm long. But it's generic SPI transmission so you can search for other topics not necessarily related to HyperHDR/HyperSPI and estimate probably maximum distance you can achieve. For sure lowering data speed transmission from default 25Mb and using shielded cables could extend it. Shielded Ethernet cable could be a good choice for longer distance.

> Are you saying that you need to prepare a separate power supply for Rpi4? And do you connect the gpio ground terminal to the SMPS minus that powers the LEDs?

Yes, for RPi4 provide separate 2A USB power supply but SOLID one. Preferable genuine one as Rpi4 is a bit picky about that. For common ground connect one of Rpi GPIO ground (there are few of them available) to minus that powers the LED strip. It's properly connected on your new diagram. for SN74AHCT125N  I've already provided you link that describes how to connect it: https://github.com/awawa-dev/HyperHDR/discussions/177

**Comment by bong8839** [comment]
The optimal length of the connecting cable between the Rpi4 and the ESP is 15cm, it can be tailored. The 1M length of the data cable can be a problem because it is too long due to voltage drop, etc., isn't it a problem that can be solved by using a level shifter? What is the appropriate length in the system without changing the data transfer rate?

Lastly, I keep asking about 74AHCT125, and I read the link you gave me and thought about it, so I drew the wiring diagram again. As I mentioned at the beginning, I'm not sure about what I understood as a beginner, so I keep asking for confirmation on the wiring diagram.

You explain well, but I'm really sorry that I didn't understand;

![image](https://user-images.githubusercontent.com/97230290/148431102-a6b80778-aa57-4885-ade5-a3f07c32abe3.png)

**Comment by awawa-dev** [comment]
No problem, it's a fairly challenging project. And it's a bit tricky if you want to get it right achieving maximum performance, without relying on your luck. 

Regarding level shifter and the long data line: that depends where you put that level shifter...if next to the LED strip then OK. But if rather next to Rpi then we have 1m data line where the 5v will drop again. Knowing your data cable resistance you can calculate it. But probably the bigger problem could be EMI interference on that long data line. So may or may not work...if I have to guess then I would say it should work with the level shifter. But without any guarantee if it works well all the time without some random flashes due to some interference..didn't test similar solution and you will have to check it out yourself in your environmental. If the distance is the problem then maybe choose HyperSerialEsp32 / Adalight solution? USB cable from Rpi to Esp could have even few meters.

On your diagram I miss common ground between Esp and Rpi: if you connect Esp to Rpi using USB cable then you will achieve it. But for better stability of SPI communication you should connect one of the Rpi ground to the Esp ground also. The rest on the diagram seems fine.
If you have troubles with soldering cables to 74AHCT125 legs then you can use: DIP socket or/and prototype board.

Edit: I've measured my data line and it's 55cm from Esp&level shifter to the LED strip. Power cables (+,-) are similar length. Maybe it will work over 1meter and you can give it a try, if won't you can switch the solution to HyperSerialEsp32.

**Reply by bong8839** [reply]
Thank you for checking up to your line length. First I ordered the parts. It will take some time for the delivery to arrive, but you should work carefully as soon as it arrives. Rpi and ESP grounding will also be carried out together! I will work on the product as soon as it arrives, and if I have any additional questions, I will leave a question again. I want to make your system clean and successful! Good day!

**Reply by herrzettpunkt** [reply]
>I will work on the product as soon as it arrives

@bong8839 Have you gotten any further? I'd like to build exactly the same setup as you. Up until now I've had my sk6812 strip running directly over the GIO18, but it flickers every now and then.

Now I would like to install a D1 Mini ESP32 and a level shifter in the hope that my 5M strip will stop flickering.

I would be particularly interested in how you programmed the ESP32.

Thanks very much!

**Reply by bong8839** [reply]
@herrzettpunkt Hi. All materials were purchased. However, I am not able to proceed with the work due to a lot of work, and I am a beginner, so I expect there will be difficulties, but I will try it soon.

I bought 8266, not esp32, and purchased the D1 mini that awawa-dev had purchased. First of all, it is expected that uploading the sketch will be done according to the contents of the link, but if the work is completed well, I will write an article. However, some difficulties are expected ^^

**Reply by bong8839** [reply]
> > 상품이 도착하는 대로 작업을 하겠습니다
> 
> @봉8839 더 알아낸 거 있어? 당신과 정확히 같은 설정을 만들고 싶습니다. 지금까지 저는 제 sk6812 스트립을 GIO18을 바로 넘어가 보았지만, 가끔씩 깜빡입니다.
> 
> 이제 제 5M 스트립이 깜빡이는 것을 멈추기를 바라는 마음으로 D1 미니 ESP32와 레벨 시프터를 설치하고 싶습니다.
> 
> 저는 특히 ESP32를 어떻게 프로그래밍했는지 관심이 있습니다.
> 
> 정말 감사합니다!

https://github.com/awawa-dev/HyperHDR/discussions/207

First of all, the hardware worked as the picture in the link below.  It works perfectly.  Any other questions?
